test/integration/ipamperf/results.go

```diff
@@ -0,0 +1,215 @@
+type NodeDuration struct {
+	Name     string       // node name
+	PodCIDR  string       // the podCIDR that was assigned to the node
+	Duration JSONDuration // how long it took to assign podCIDR
+}
+
+// Results represents the observed test results.
+type Results struct {
+	Name           string         // name for the test
+	Config         *Config        // handle to the test config
+	Succeeded      bool           // whether all nodes were assigned podCIDR
+	MaxAllocTime   JSONDuration   // the maximum time take for assignment per node
+	TotalAllocTime JSONDuration   // duration between first addition and last assignment
+	NodeAllocTime  []NodeDuration // assignment time by node name
+}
+
+// NewObserver creates a new observer given a handle to the Clientset
+func NewObserver(clientSet *clientset.Clientset, numNodes int) *Observer {
+	o := &Observer{
+		timing:    map[string]*nodeTime{},
+		numNodes:  numNodes,
+		clientSet: clientSet,
+	}
+	return o
+}
+
+// StartObserving starts an asynchronous loop to monitor for node changes.
+// Call Results() to get the test results after starting observer.
+func (o *Observer) StartObserving() (util.ShutdownFunc, error) {
+	w, err := o.clientSet.CoreV1().Nodes().Watch(metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	go o.monitor(w)
+	glog.Infof("Test observer started")
+	stopFunc := func() {
+		if w != nil {
+			w.Stop()
+		}
+	}
+	return stopFunc, nil
+}
+
+// Results returns the test results. It waits for the observer to finish
+// and returns the computed results of the observations.
+func (o *Observer) Results(name string, config *Config) *Results {
+	var (
+		firstAdd       = time.Now()
+		lastAssignment time.Time
+	)
+	o.wg.Wait()
+	results := &Results{
+		Name:          name,
+		Config:        config,
+		Succeeded:     o.numAdded == o.numNodes && o.numAllocated == o.numNodes,
+		MaxAllocTime:  0,
+		NodeAllocTime: []NodeDuration{},
+	}
+	for name, nTime := range o.timing {
+		addFound := !nTime.added.IsZero()
+		if addFound && nTime.added.Before(firstAdd) {
+			firstAdd = nTime.added
+		}
+		cidrFound := !nTime.allocated.IsZero()
+		if cidrFound && nTime.allocated.After(lastAssignment) {
+			lastAssignment = nTime.allocated
+		}
+		if addFound && cidrFound {
+			allocTime := nTime.allocated.Sub(nTime.added)
+			if allocTime > time.Duration(results.MaxAllocTime) {
+				results.MaxAllocTime = JSONDuration(allocTime)
+			}
+			results.NodeAllocTime = append(results.NodeAllocTime, NodeDuration{
+				Name: name, PodCIDR: nTime.podCIDR, Duration: JSONDuration(allocTime),
+			})
+		}
+	}
+	results.TotalAllocTime = JSONDuration(lastAssignment.Sub(firstAdd))
+	sort.Slice(results.NodeAllocTime, func(i, j int) bool {
+		return results.NodeAllocTime[i].Duration > results.NodeAllocTime[j].Duration
+	})
+	return results
+}
+
+func (o *Observer) monitor(w watch.Interface) {
+	o.wg.Add(1)
+	defer o.wg.Done()
+
+	for event := range w.ResultChan() {
```

Comment:
Instead of manually picking up the watch events and processing them, you might want to use an informer with attached event-handler functions. E.g - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/scalability/density.go#L681

That machinery IIUC offers more robustness compared to natively reading watch events (e.g if watch closed or timed out). cc @wojtek-t to confirm (who knows better about this).

Reply:
Yes definitely.
Single watch isn't very robust - it may be broken due to network issue, some issue on apiserver, etc. And you don't want to fail the whole test because of that.